HSEARCH-2615 Embedded ID is not well handled

mincong-h opened the pull request:
Hi @yrodiere ,

Here's the PR for <https://hibernate.atlassian.net/browse/HSEARCH-2615>. I actually didn't finish the ticket, but I want to have your opinion about this ticket before going further. The origin of this issue is that for entity-types having an `@EmbbededId`, their order is not well handled. For example, given an `@Embeddable` id having 3 fields (year, month, day-of-month), when indexing entities >= (2017, 06, 20), then (2017, 07, 01) will not be indexed, because the the field day-of-month 01 < 20.

However, this is not always true. After adding 2 test-cases in `MassIndexingJobWithEmbeddedIdTest`, I found that:

- `Comparable` embedded ID are handled correctly
- Non `Comparable` embedded ID are not handled correctly

It means that we have 2 solutions:

1. Restrict `@Embeddable` ID to be comparable, and document it
2. Allow `@Embeddable` ID to be non-comparable, and discover "nested" properties in the ID, if any, and build a sort + criteria for partitions explicitly based on that metadata, as you mentioned in the last exchange.

Which approach do you prefer?

mincong-h committed:
HSEARCH-2615 Handle sorting and partitioning on @IdClass & @EmbeddedId

yrodiere commented:
@mincong-h Thanks! But actually your test, as it is, is invalid. The `ComparableDateId` type only has one attribute (`LocalDate d`), and it's using field access mode (see http://docs.jboss.org/hibernate/orm/5.2/userguide/html_single/Hibernate_User_Guide.html#access), so Hibernate ORM actually only created one column in database and used this column to perform comparison. In which case the problem will obviously never show up, since the ID is not composite after all.

So I still think solution 2 is our only way out: the comparison will be performed in database, so the composite ID being `Comparable` or not does not matter.

mincong-h committed:
HSEARCH-2615 Test sorting and partitioning on @IdClass & @EmbeddedId

mincong-h commented:
Hi @yrodiere, thanks for your hint! Here's the new implementation.

- The new implementation covers both `@IdClass` and `@EmbeddedId`, but using different approach.
- From now on, the ID ordering is applied to all the scrollable results in partition-mapping and entity-reading. Their implementation varies based on the ID type (`@Id`, `@IdClass`, `@EmbeddedId`).
- The checkpoint-ID is now considered as part of the partition boundary, which facilitates the ID processing: we only need to take care lower ID and upper ID.
- ID processing is delegated to enum `IdProcessor`.

yrodiere commented:
@mincong-h Thanks! I'll try to have a look by the end of the week.

yrodiere commented:
@mincong-h I took the liberty of rebasing and squashing your work, then force-pushing it to your branch, so as to avoid unnecessary discussions on back-and-forth changes that are not present after the last commit.
Your original PR is still available here: https://github.com/yrodiere/hibernate-search/tree/HSEARCH-2615_beforeSquashing

Starting the review now :)

yrodiere commented on jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/PersistenceUtil.java:
```diff
@@ -71,4 +221,127 @@ public static StatelessSession openStatelessSession(EntityManagerFactory entityM
+				return orders;
+
+			default:
+				throw new IllegalStateException( "Cannot determine IdType: this should never happen." );
```
This should be an AssertionFailure.

mincong-h replied:
You're right. By doing using the `org.hibernate.search.exception.AssertionFailure`, we clarify the responsibility of the exception: it is an internal error, and users should report a bug if they meet it. This is also written in the Javadoc of `AssertionFailure`:

> Indicates failure of an assertion: a possible bug in Hibernate Search.

yrodiere commented on jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/PersistenceUtil.java:
```diff
@@ -71,4 +221,127 @@ public static StatelessSession openStatelessSession(EntityManagerFactory entityM
+
+			case ID_CLASS:
+				attributeList = new ArrayList<>( entityType.getIdClassAttributes() );
+				attributeList.sort( Comparator.comparing( Attribute::getName ) );
```
We'll probably want to move that to some job-level resource, like the entity name => entity class map. Then we will only retrieve and sort the attributes list once.

yrodiere commented on jsr352/core/src/test/java/org/hibernate/search/jsr352/massindexing/MassIndexingJobWithCompositeIdTest.java:
```diff
@@ -0,0 +1,159 @@
+	 * of our customized solution: we use our own comparator—a lexicographical
+	 * order of attribute's name—for ID attributes sorting. We assume that this
+	 * order is always true during the entire indexing process, but it can't be
+	 * when user adds his own restriction on the ID field.
```
To be fair, it's not really a limitation of *our* solution. It's just that the `Restrictions.ge` below generates this condition:

```
year >= 2017 AND month >= 6 AND day >= 20
```

... which is obviously not what any sane person would expect, since this will exclude July 1st to July 19th.
But still, the end result is consistent with the restriction provided by the user: we only index those documents that match the given criterion, and we index all documents that match the given criterion.

Were the restriction to behave correctly, it should also work well with our solution.

yrodiere commented on jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/PersistenceUtil.java:
```diff
@@ -71,4 +221,127 @@ public static StatelessSession openStatelessSession(EntityManagerFactory entityM
+
+	private static Object getProperty(Object obj, String propertyName)
+			throws IntrospectionException, InvocationTargetException, IllegalAccessException {
+		return new PropertyDescriptor( propertyName, obj.getClass() ).getReadMethod().invoke( obj );
```
This is really not a good idea, because we may end up getting the value in a different way than Hibernate ORM's. I'll try to see if there's a way to get a value from a singular attribute and the ID...

yrodiere committed:
HSEARCH-2615 Cache Hibernate ORM metadata (ID order etc.) as much as possible

So that we'll avoid checking out the metadata for every single query.

This cache will be most effective on JSR-352 runtimes that actually share
the job context between partitions, such as JBeret; others such as
JBatch force us to rebuild the context for each partition, in which case
the only benefic will be to only rebuild the metadata once for both
the "greater-or-equal" and "lesser" restrictions.

mincong-h commented:
> I took the liberty of rebasing and squashing your work, ...

I should have done before the PR. I wanted to keep the commit history at the beginning, but that was too hard to achieve... Thank you! And thanks for keeping the before-squash branch too. I think we won't need it, you can delete it.

yrodiere committed:
HSEARCH-2615 Use Hibernate ORM APIs/SPIs to retrieve ID property values

This should avoid problems in particular cases where some property names
are ambiguous if we don't take Hibernate ORM's access mode into account.

yrodiere committed:
HSEARCH-2615 Clarify that Restrictions.ge et al. don't work well on composite IDs

yrodiere committed:
HSEARCH-2615 Ensure there's always more than 1 partition in MassIndexingJobWithCompositeIdTest

yrodiere committed:
HSEARCH-2615 Fix the greater-or-equal operator for CompositeIdOrder

With PX as the property names, and VX as those properties' values
in the compared ID, we used to translate the greater-or-equal operator
as P1 >= V1 OR P1 = V1 AND P2 >= V2,
but that's absurd as it's equivalent to P1 >= V1 (the first clause
will always be true if the second one is true).

The actual condition we were looking for was
P1 > V1 OR P1 = V1 AND P2 > V2 OR P1 = V1 AND P2 = V2.

This commit also slightly changes the tests in order to
reproduce the bug, which used to happen in
MassIndexingJobWithCompositeIdTest.canHandleIdClass_strategyFull()
for instance, where partitions 2 ([2017-06-14,2017-07-20[) and
3 ([2017-07-20,2017-06-27[) both processed the date 2017-06-20
(with the lexicographical order ("day", "month", "year").

yrodiere reviewed (approved):
Looks good! I added a few commits, but only the second and last one actually have a functional impact (they fix bugs), the rest is merely caching/cosmetics.

yrodiere commented:
Merged, thanks!

mincong-h commented on jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/CompositeIdOrder.java:
```diff
@@ -113,10 +119,21 @@ private Criterion restrictLexicographically(BiFunction<String, Object, SimpleExp
 			// The last expression has whatever symbol is defined by "lastRestriction"
```
My trivial comment: should we update the `"lastRestriction"` here? The variable is renamed to `"strictOperator"` now

yrodiere replied:
Right, thanks for pointing this out. I pushed a commit directly to the jsr352 branch to fix this and another typo.

mincong-h commented:
@yrodiere , I really appreciate your review on my PR and I learned a lot from it:

- High cohesion
- Low coupling
- Variable naming
- Java 8 usage
- Collections API handling (initial capacity, choice of interface type)
- Immutability
- Hibernate API usage
- Algorithm fix

That's really awesome. Thanks for keeping the commit history as well. :+1:
